flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaRecordSerializationSchema.java

```diff
@@ -144,14 +147,34 @@ public void open(
         valueSerialization.open(context);
     }
 
+    private String getTargetTopic(RowData element) {
+        final String topic = readMetadata(element, KafkaDynamicSink.WritableMetadata.TOPIC);
+        if (topic == null && topics == null) {
+            throw new IllegalArgumentException(
+                    "The topic of the sink record is not valid. Expected a single topic but no topic is set.");
+        } else if (topic == null && topics.size() == 1) {
+            return topics.get(0);
+        } else if (topics != null && topics.size() > 0 && !topics.contains(topic)) {
```

Comment:
Open to discuss: I decided to keep topics as a List since, but we can change it to HashSet. I am assuming the list will be short on average.

List can outperform HashSet for small number of elements:
https://stackoverflow.com/questions/150750/hashset-vs-list-performance